File: Calibration.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <math.h>
#include <stdlib.h>  
#include <random>

using namespace std;
using namespace cv;

double cal_normal_random(double mean, double sigma)
{
	double rand_num[12], r, r2, c1, c2, c3, c4, c5, gauss_rand;
	unsigned long max;
	int i;

	max = 2147483647;
	for (i = 0; i < 12; i++) {
		rand_num[i] = ((double) random()) / ((double)(max));
	}
	c1 = 0.029899776;
	c2 = 0.008355968;
	c3 = 0.076542912;
	c4 = 0.252408784;
	c5 = 3.949846138;

	r = 0.0;
	for (i = 0; i < 12; i++)
		r += rand_num[i];
	r = (r - 6.0) / 4.0;
	r2 = r * r;
	gauss_rand = ((((c1 * r2 + c2) * r2 + c3) * r2 + c4) * r2 + c5) * r;
	return(mean + sigma * gauss_rand);
}

void CrossPoints(std::vector<cv::Point3d> Gaze, std::vector<cv::Point2d> &ImagePoints) {

	for (auto i = Gaze.begin(); i != Gaze.end(); i++) {
		double x = (1 / i->z) * i->x;
		double y = (1 / i->z) * i->y;
		ImagePoints.push_back(cv::Point2d(x, y));
	}
}

void add_noise_gaze(std::vector<cv::Point3d> GazeVector, int k, std::vector<cv::Point3d> &GazeVector_noise) {

	for (auto i = GazeVector.begin(); i != GazeVector.end(); i++) {
		double mag = sqrt(pow(i->x, 2) + pow(i->y, 2) + pow(i->z, 2));
		double random = cal_normal_random(0, (k+1) * 0.01 * mag);
		//cout << "random = " << random << endl;
		GazeVector_noise.push_back(cv::Point3d(i->x + random, i->y + random, i->z + random));
	}
}

void unknown_scale(cv::Mat rot_Mat, cv::Mat tvecs, std::vector<cv::Point3d> WorldCoord, cv::Mat camera_matrix, std::vector<double> &Scale) {

	for (auto i = WorldCoord.begin(); i != WorldCoord.end(); i++) {
		cv::Mat Sample_3Dpoint = (cv::Mat_<double>(3, 1) << i->x, i->y, i->z);
		cv::Mat Rs = (rot_Mat * Sample_3Dpoint) + tvecs;
		cv::Mat Pixel_coord = camera_matrix * Rs;
		Scale.push_back(Pixel_coord.at<double>(2));
	}

}

void Reprojection(std::vector<cv::Point2d> ImagePoints_noise, std::vector<double> Scale, cv::Mat rot_Mat, cv::Mat tvecs, cv::Mat camera_matrix, cv::Mat &ReprojectionPoints) {
	
	cv::Mat ImagePoints_noise_conc;// = cv::Mat::zeros(6, 3, CV_64F);
	cv::Mat tmpMat1 = cv::Mat(ImagePoints_noise).reshape(1); // converting vector to Matrix
	cv::Mat tmpMat2 = cv::Mat::ones(6, 1, CV_64F);
	hconcat(tmpMat1, tmpMat2, ImagePoints_noise_conc);
	cv::Mat first_part = rot_Mat.inv() * camera_matrix.inv();
	cv::Mat third_part = rot_Mat.inv() * tvecs;
	int counter = 0;
	for (auto i = Scale.begin(); i != Scale.end(); i++) {
		cv::Mat temp = *i * ImagePoints_noise_conc.row(counter); 
		cv::Mat secon_part = first_part * (temp.t());
		ReprojectionPoints.row(counter) = (secon_part - third_part).t();
		counter++;
	}

}

void Reprojection_error(cv::Mat ReprojectionPoints, std::vector<cv::Point3d> WorldCoord, std::vector<double> &error) {
	
	std::vector<double> diff;
	cv::Mat WorldCoordMat = cv::Mat(WorldCoord).reshape(1);
	for (int i = 0; i < 6; i++) {
		diff.push_back(cv::norm(ReprojectionPoints.row(i) - WorldCoordMat.row(i)));
	}
	float average = accumulate(diff.begin(), diff.end(), 0.0) / diff.size();
	error.push_back(average);
}

int main(int argc, char** argv)
{
	FILE* Output;
	Output = fopen("Error.txt", "w");

	std::vector<cv::Point3d> WorldCoord;
	WorldCoord.push_back(cv::Point3d(0, 3, 50));
	WorldCoord.push_back(cv::Point3d(2, -5, 47));
	WorldCoord.push_back(cv::Point3d(-1, 7, 60));
	WorldCoord.push_back(cv::Point3d(5, -1, 40));
	WorldCoord.push_back(cv::Point3d(0, 2, 45));
	WorldCoord.push_back(cv::Point3d(3, -4, 44));

	std::vector<cv::Point3d> GazeVector;
	GazeVector.push_back(cv::Point3d(0, 3, 30));
	GazeVector.push_back(cv::Point3d(-3, -5, 28));
	GazeVector.push_back(cv::Point3d(10, 7, 31));
	GazeVector.push_back(cv::Point3d(-10, 1, 25));
	GazeVector.push_back(cv::Point3d(-5, 2, 30));
	GazeVector.push_back(cv::Point3d(-6, -4, 27));

	cv::Mat camera_matrix = cv::Mat::eye(3, 3, CV_64F);
	std::vector<cv::Point2d> ImagePoints;
	std::vector<double> dist_coeffs(4,0);
	std::vector<double> error_PnP;
	std::vector<double> error_PnPRansac;
	for (int k = -1; k < 20; k++) {
		std::vector<cv::Point2d> ImagePoints_noise;
		std::vector<cv::Point3d> GazeVector_noise;

		cv::Mat rvecs_PnP = cv::Mat::zeros(3, 1, CV_64FC1);
		cv::Mat tvecs_PnP = cv::Mat::zeros(3, 1, CV_64FC1);
		std::vector<double> Scale_PnP;
		cv::Mat rot_Mat_PnP = cv::Mat::zeros(3, 3, CV_64FC1);
		cv::Mat ReprojectionPoints_PnP = cv::Mat::zeros(cv::Size(3, 6), CV_64FC1);

		cv::Mat rvecs_PnPRansac = cv::Mat::zeros(3, 1, CV_64FC1);
		cv::Mat tvecs_PnPRansac = cv::Mat::zeros(3, 1, CV_64FC1);
		std::vector<double> Scale_PnPRansac;
		cv::Mat rot_Mat_PnPRansac = cv::Mat::zeros(3, 3, CV_64FC1);
		cv::Mat ReprojectionPoints_PnPRansac = cv::Mat::zeros(cv::Size(3, 6), CV_64FC1);

		add_noise_gaze(GazeVector, k,  GazeVector_noise);															// Output: GazeVector_noise
		CrossPoints(GazeVector_noise, ImagePoints_noise);															// Output: ImagePoints_noise

		cv::solvePnP(WorldCoord, ImagePoints_noise, camera_matrix, dist_coeffs, rvecs_PnP, tvecs_PnP, CV_EPNP);			// Outputs: rvecs, tvecs
		cv::solvePnPRansac(WorldCoord, ImagePoints_noise, camera_matrix, dist_coeffs, rvecs_PnPRansac, tvecs_PnPRansac);
		cv::Rodrigues(rvecs_PnP, rot_Mat_PnP);																			//Output: rot_Mat
		cv::Rodrigues(rvecs_PnPRansac, rot_Mat_PnPRansac);
		unknown_scale(rot_Mat_PnP, tvecs_PnP, WorldCoord, camera_matrix, Scale_PnP);									// Output: Scale
		unknown_scale(rot_Mat_PnPRansac, tvecs_PnPRansac, WorldCoord, camera_matrix, Scale_PnPRansac);
		Reprojection(ImagePoints_noise, Scale_PnP, rot_Mat_PnP, tvecs_PnP, camera_matrix, ReprojectionPoints_PnP);		// Output: ReprojectionPoints
		Reprojection(ImagePoints_noise, Scale_PnPRansac, rot_Mat_PnPRansac, tvecs_PnPRansac, camera_matrix, ReprojectionPoints_PnPRansac);
		Reprojection_error(ReprojectionPoints_PnP, WorldCoord, error_PnP);												// Output: error
		Reprojection_error(ReprojectionPoints_PnPRansac, WorldCoord, error_PnPRansac);
		if (k == -1) {
			cout << "RotationMatrix:" << endl << rot_Mat_PnP << endl;
			cout << "TranslationMatrix:" << endl << tvecs_PnP << endl;
		}

		//cout << error_PnP.at(k) << "\t" << error_PnPRansac.at(k) << endl;
	}

	fprintf(Output,"PnP PnPRansac\n");
	for (int i = 0; i < 20; i++)
		fprintf(Output, "%f %f\n", error_PnP.at(i), error_PnPRansac.at(i)); // Write to file

	fclose(Output);
	return 0;

}

File: PythonCodes/Calibration.py
import numpy as np
import cv2
import random
import matplotlib.pyplot as plt

class Calibration(object):
    
    def __init__(self,GazeVector, WorldCoord):
        '''
        Initilization of GazeVector and World Coordinate and camera matrix
        '''
        self.GazeVector = GazeVector
        self.WorldCoord = WorldCoord
        self.camera_matrix = np.array([ [1, 0, 0],
                           [0, 1, 0],
                           [0, 0, 1]], dtype = "double")
        self.dist_coeffs = np.zeros((4,1)) # Assuming no lens distortion
        
    def CrossSectionPoints(self, GazeVector):
        '''
        Compute intersection of whole gaze vectors with the virtual plane and return numpy array
        '''
        ImagePoints = np.zeros((GazeVector.shape[0],2),dtype="double")
        for i in range(GazeVector.shape[0]):
            ImagePoints[i,:] = self.CrossSectionPoint(GazeVector[i,:])
        
#        print("ImagePoints", ImagePoints)
        return ImagePoints
    
    def CrossSectionPoint(self,GazeVec):
        '''
        Computer intersection of only one gaze vector and return to a single point (x,y)
        '''
        t= 1/GazeVec[2]
        X_sec = GazeVec[0] * t
        Y_sec = GazeVec[1] * t
        return X_sec,Y_sec
    
    def PnP(self, WorldCoord,ImagePoints):
        '''
        Apply PnP routine and return rotation and translation matrices
        '''
        (_,rvecs, tvecs) = cv2.solvePnP(WorldCoord,ImagePoints, self.camera_matrix, 
                                        self.dist_coeffs, flags=cv2.SOLVEPNP_EPNP)
        Rt, _ = cv2.Rodrigues(rvecs)
        return Rt, tvecs
    
    def PnPRansac(self, WorldCoord,ImagePoints):
        '''
        Apply PnPRansac routine and return rotation and translation matrices
        PnPRansac is less sensitive to errors
        '''
        (_, rvecs, tvecs, inliers) = cv2.solvePnPRansac(WorldCoord,ImagePoints, self.camera_matrix, 
                                                        self.dist_coeffs)
        Rt, _ = cv2.Rodrigues(rvecs)
        print("tvecs= ", tvecs)
        print("Rt=", Rt)
        return Rt, tvecs
    
    def cal_normal_random(self,mean,sigma):
        '''
        Calculate Gaussian noise to be added to the gaze vector elements
        '''
        rand_num = np.zeros(12)
        max_num = 2147483647 
        
        for i in range(12):
            rand_num[i] = random.randint(0, 2147483647)/max_num
    
        c1 = 0.029899776 ;
        c2 = 0.008355968 ;
        c3 = 0.076542912 ;
        c4 = 0.252408784 ;
        c5 = 3.949846138 ;
        
        r = 0.0 ;
        for i in range(12):
            r+= rand_num[i] ;
        
        r = (r-6.0)/4.0 ;
        r2 = r*r ;
        gauss_rand=((((c1*r2+c2)*r2+c3)*r2+c4)*r2+c5)*r
        
        return ((mean+sigma*gauss_rand))

    def Get_ScaleFactor(self,Rt,translation_vector):
        '''
        Compute the unknown scale factor 
        '''
        
        scalefactor = np.zeros((self.WorldCoord.shape[0],1))
        for i in range(self.WorldCoord.shape[0]):
            Sample_3Dpoint = self.WorldCoord[i].reshape(3,1)
            R_s = np.dot(Rt,Sample_3Dpoint).reshape(3,1) + translation_vector.reshape(3,1)
            Pixel_coord = np.dot(self.camera_matrix,R_s)
            scalefactor[i] = Pixel_coord[2]
#            print(scalefactor[i])
        return scalefactor
    
    def Reprojection(self,ImagePoints,S,Rt,translation_vector):
        '''
        Reproject Image points into 3D world coordinate systems to measure error
        '''
        ReprojectionPoints = np.zeros((6,3))
        ImagePoints_conc = np.concatenate((ImagePoints, np.array([1] * 6).reshape(6,1) ), axis=1) 
        ImagePoints_scale = np.multiply(ImagePoints_conc,S)
        
        for i in range(6):
            first_part = np.dot(np.linalg.inv(Rt),np.linalg.inv(self.camera_matrix))
            secon_part = np.dot(first_part, ImagePoints_scale[i])
            third_part = np.dot(np.linalg.inv(Rt),translation_vector)
            ReprojectionPoints[i,:] =  np.transpose((secon_part.reshape(3,1) - third_part))
            
        return ReprojectionPoints
    
    def ImageNoise(self, k):
        '''
        Add noise to images
        '''
        GazeVector_noise = np.zeros((6, 3))
        for i in range(self.GazeVector.shape[0]):
            m = np.linalg.norm(self.GazeVector[i])
            random_quantity = self.cal_normal_random(0,(k+1)*0.01*m)
            print("random=", random_quantity)
            GazeVector_noise[i,:] = (self.GazeVector[i] + random_quantity).reshape(1,3)
            
        ImagePoints_noise = self.CrossSectionPoints(GazeVector_noise) 
        return ImagePoints_noise 
    
    def error(self,ImagePoints_noise,Rt,translation_vector):
        '''
        Compute the reprojection error
        '''
        S = self.Get_ScaleFactor(Rt,translation_vector) 
        ReprojectionPoints = self.Reprojection(ImagePoints_noise, S,Rt,translation_vector)
        reprojection_error = self.WorldCoord - ReprojectionPoints
        
        return reprojection_error
        
    def run(self,Number_Iter):
        '''
         Measure the error as sum of square root of the difference 
         between 3D projected and noisless 3D points
        '''
        error_PnP = np.zeros(Number_Iter)
        error_PnPRansac = np.zeros(Number_Iter)
        
        for k in range(Number_Iter):

            ImagePoints_noise = self.ImageNoise(k)
            Rt_pnp,translation_vector_pnp = self.PnP(self.WorldCoord,ImagePoints_noise)
            Rt_Ransac,translation_vector_Ransac = self.PnPRansac(self.WorldCoord,ImagePoints_noise)
            PnP_diff = self.error(ImagePoints_noise,Rt_pnp,translation_vector_pnp)
            Ransac_diff = self.error(ImagePoints_noise,Rt_Ransac,translation_vector_Ransac)
            error_PnP[k] = np.average(np.linalg.norm(PnP_diff, axis=1))
            error_PnPRansac[k] = np.average(np.linalg.norm(Ransac_diff, axis=1))
        
        return error_PnP,error_PnPRansac
    

        